artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java

```diff
@@ -151,25 +171,20 @@ public synchronized void addTail(final MessageReference ref, final boolean direc
          HolderReference hr = map.get(prop);
 
          if (hr != null) {
-            // We need to overwrite the old ref with the new one and ack the old one
-
-            replaceLVQMessage(ref, hr);
-
-            if (isNonDestructive() && hr.isDelivered()) {
-               hr.resetDelivered();
-               // --------------------------------------------------------------------------------
-               // If non Destructive, and if a reference was previously delivered
-               // we would not be able to receive this message again
-               // unless we reset the iterators
-               // The message is not removed, so we can't actually remove it
-               // a result of this operation is that previously delivered messages
-               // will probably be delivered again.
-               // if we ever want to avoid other redeliveries we would have to implement a reset or redeliver
-               // operation on the iterator for a single message
-               resetAllIterators();
-               deliverAsync();
-            }
+            if (isNonDestructive() && hr.isInDelivery()) {
+               // if the ref is already being delivered we'll do the replace in the postAcknowledge
+               hr.setReplacementRef(ref);
+            } else {
+               // We need to overwrite the old ref with the new one and ack the old one
+               replaceLVQMessage(ref, hr);
 
+               if (isNonDestructive() && hr.isDelivered()) {
+                  hr.resetDelivered();
+                  // since we're replacing a ref that was already delivered we want to trigger a "special" delivery for this new replacement
```

Comment:
I - stale comment... the "special" word...

ii - just an idea... what about rename repeatNextDelivery(messageReference) as nextDelivery(messageReference)

and call it here instead of using the collection directly? 

just a suggestion!

Reply:
I fixed "special" in the comment.

I didn't change `repeatNextDelivery`. I used "repeat" in the name because it serves the same purpose as `repeat` on the `LinkedListIterator` in `QueueImpl`. I wanted that to be clear so its easy to see that if the consumer is "busy" we just repeat the delivery no matter what.